collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormsDetailsFetcher.kt

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2018 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.odk.collect.android.formmanagement
+
+import org.odk.collect.analytics.Analytics.Companion.log
+import org.odk.collect.android.analytics.AnalyticsEvents
+import org.odk.collect.android.openrosa.OpenRosaFormSource
+import org.odk.collect.android.utilities.FormUtils
+import org.odk.collect.android.utilities.WebCredentialsUtils
+import org.odk.collect.forms.Form
+import org.odk.collect.forms.FormSource
+import org.odk.collect.forms.FormSourceException
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.ManifestFile
+import org.odk.collect.forms.MediaFile
+import org.odk.collect.shared.strings.Md5.getMd5Hash
+import timber.log.Timber
+import java.io.File
+
+/**
+ * Open to allow mocking (used in existing Java tests)
+ */
+open class ServerFormsDetailsFetcher(
+    private val formsRepository: FormsRepository,
+    private val formSource: FormSource,
+    private val diskFormsSynchronizer: DiskFormsSynchronizer
+) {
+    open fun updateUrl(url: String) {
+        (formSource as OpenRosaFormSource).updateUrl(url)
+    }
+
+    open fun updateCredentials(webCredentialsUtils: WebCredentialsUtils) {
+        (formSource as OpenRosaFormSource).updateWebCredentialsUtils(webCredentialsUtils)
+    }
+
+    @Throws(FormSourceException::class)
+    open fun fetchFormDetails(): List<ServerFormDetails> {
+        diskFormsSynchronizer.synchronize()
+
+        val formList = formSource.fetchFormList()
+        return formList.map { listItem ->
+            val manifestFile = listItem.manifestURL.let {
```

Comment:
I think you can simplify this:
```kotlin
val manifestFile = listItem.manifestURL?.let {
    getManifestFile(formSource, it)
}
```

Reply:
Yes! I remember writing this and being like "I swear there's a cleaner way to do this I'm forgetting". `?.let` is a great trick - basically `map` from a `Maybe` monad.